File: src/main/java/ml/FPGrowthMethods.java
package ml;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.ml.fpm.FPGrowth;
import org.apache.spark.ml.fpm.FPGrowthModel;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.*;
import java.util.*;
import java.util.stream.Collectors;


public class FPGrowthMethods {
    public static void main(String[] args) {

        SparkSession spark = SparkSession
                .builder()
                .appName("FPGrowthMethods")
                .master("local[2]")
                .getOrCreate();

        spark.sparkContext().setLogLevel("ERROR");

        Dataset<Row> topicMethodsDF = spark.read().parquet("hdfs://ip243:8020/zsy/warehouse/ods/tiku/topic_methods");

        JavaRDD<Row> methodIdsRDD = topicMethodsDF
                .select("method_ids")
                .toJavaRDD()
                //.filter(row -> row.getString(0).contains(","))
                .filter(row ->{
                    Set<Integer> set = new HashSet<>();
                    String[] split = row.getString(0).split("[,]");
                    for (String s : split) {
                        set.add(Integer.parseInt(s));
                    }
                    return set.size() == split.length;
                });


        final List<Row> dataMethods =  methodIdsRDD.collect().stream().map(row -> RowFactory.create(Arrays.asList(row.getString(0).split("[,]")))).collect(Collectors.toList());

        StructType schema = new StructType(new StructField[]{ new StructField(
                "items", new ArrayType(DataTypes.StringType, true), false, Metadata.empty())
        });

        Dataset<Row> itemsDF  = spark.createDataFrame(dataMethods, schema);

        final FPGrowthModel model = new FPGrowth()
                .setItemsCol("items")
                .setMinSupport(0.0001)
                .setMinConfidence(0.7)
                .fit(itemsDF);

        /**
         * Display frequent itemsets.
         */

        //model.freqItemsets().printSchema();
        //System.out.println("freqItemsets--->");
        //model.freqItemsets().show();
        System.out.println("-----频繁数据集-----");
        System.out.println("-----考点ids, 出现的频次-----");
        int methodSum = 0;
        for (Row row : model.freqItemsets().toJavaRDD().collect()) {
            List<String> list = row.getList(0);
            if (list.size() > 1) {
                methodSum++;
            }
            long count = row.getLong(1);
            System.out.println(list.toString()+"--->"+count);
        }
        System.out.println("methodSum---->"+methodSum);

        /**
         * Display generated association rules.
         */
        //model.associationRules().printSchema();
        //System.out.println("associationRules--->");
        //model.associationRules().show();

        System.out.println("-----有较强关联关系的考点-----------");
        System.out.println("-----a考点集，b考点，在考题topic中a出现后b出现的概率-----------");
        for (Row row : model.associationRules().toJavaRDD().collect()) {
            List<String> antecedentList = row.getList(0);
            List<String> consequentList = row.getList(1);
            double confidence = row.getDouble(2);
            System.out.println(antecedentList.toString()+"===>"+consequentList.toString()+", "+confidence);
        }

        System.out.println("associationRules.size--->"+model.associationRules().toJavaRDD().collect().size());

        /**
         * transform examines the input items against all the association rules and summarize the
         * consequents as prediction
         */
        //model.transform(itemsDF).show();

        spark.stop();
    }
}

File: src/main/scala/org/scala/spark/ml/FPGrowthMethods.scala
package org.scala.spark.ml

import java.util.Properties
import org.apache.spark.ml.fpm.FPGrowth
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.functions.{col, explode}
import org.apache.spark.sql.types.{LongType, StructField, StructType}
import org.apache.spark.sql._

/**
 * yqq 2020/11/17
 * 考点关联分析
 */
object FPGrowthMethods {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("FPGrowthMethods")
      .getOrCreate()

    val connectionProperties = new Properties()
    connectionProperties.put("user", "zsy")
    connectionProperties.put("password", "lc12345")

    val url = "jdbc:mysql://192.168.1.210:3307/data_mining?useUnicode=true&characterEncoding=UTF-8&autoReconnect=true&rewriteBatchedStatements=true"

    spark.sparkContext.setLogLevel("ERROR")

    import spark.implicits._

    val topicMethodsDF = spark.read.parquet("hdfs://ip243:8020/zsy/warehouse/ods/tiku/topic_methods")

    val dataFrame: DataFrame = topicMethodsDF.select("method_ids").map(row => row.getString(0).split("[,]")).toDF("items")

    val fPGrowth = new FPGrowth().setItemsCol("items").setMinSupport(0.0001).setMinConfidence(0.7)

    val model = fPGrowth.fit(dataFrame)

    val itemsets = model.freqItemsets

    /**
     * 关联考点
     */
    val relevanceMethodDF: Dataset[Row] = itemsets.filter(row => row.getList(0).size() > 1)

    val schema: StructType = relevanceMethodDF.schema add(StructField("association_id",LongType))

    val indexRDD: RDD[(Row, Long)] = relevanceMethodDF.rdd.zipWithIndex()

    val rowRDD = indexRDD.map(r => Row.merge(r._1, Row(r._2)))

    val relevanceMethod2DF = spark.createDataFrame(rowRDD, schema)

    relevanceMethod2DF.withColumn("items",explode(col("items"))).createOrReplaceTempView("association_method")
    spark.sql("select items as method_id,cast(freq as int) as association_num,cast(association_id as int) from association_method").createOrReplaceTempView("association_method")

    /**********************************************************insert to mysql*********************************************************************************/

    val methodDF = spark.read
      .format("jdbc")
      .option("url", "jdbc:mysql://192.168.1.210:3307/data_mining?useUnicode=true&characterEncoding=UTF-8&autoReconnect=true&rewriteBatchedStatements=true")
      .option("dbtable", "special_solve_method")
      .option("user", "zsy")
      .option("password", "lc12345")
      .load();

    //topicMethodDF.write().mode(SaveMode.Overwrite).parquet("hdfs://ip243:8020/zsy/warehouse/ods/tiku/topic_method");

    methodDF.createOrReplaceTempView("method");

    spark.sql("select ssm_id as method_id ,ssm_content as method_name  from method ").createOrReplaceTempView("method_name");

    val dataframe = spark.sql(" select association_method.*,method_name.method_name from association_method left join method_name on association_method.method_id = method_name.method_id")

    /**
     * 关联考点统计分析
     */
    dataframe.write
      .mode(SaveMode.Overwrite)
      .option("truncate","true")
      .jdbc(url, "association_method", connectionProperties)

    /**
     * 考点在题库出现的频次统计
     */
    val methodTopicStatDF: Dataset[Row] = itemsets.filter(row => row.getList(0).size() == 1)
    methodTopicStatDF.withColumn("items",explode(col("items"))).createOrReplaceTempView("method_stat")

    val resDF = spark.sql("select items as method_id,cast(freq as int) from method_stat")
    resDF.write
      .mode(SaveMode.Overwrite)
      .option("truncate","true")
      .jdbc(url, "method_freq_stat", connectionProperties)

    /**
     * 关联规则条件概率统计
     */
    //model.associationRules.printSchema()
    model.associationRules.createOrReplaceTempView("association_rules")
    val associationRulesDF = spark.sql("select concat_ws(',',antecedent) as antecedent,concat_ws('',consequent) as consequent,round(confidence,3) as confidence,round(lift,3) as lift from association_rules")
    associationRulesDF.write
      .mode(SaveMode.Overwrite)
      .option("truncate","true")
      .jdbc(url, "association_rules", connectionProperties)

    spark.stop()
  }
}
